plugins/reach/plugins/OpenCalais/batch/KReachVendorTaskOpenCalaisProcessorEngine.php

```diff
@@ -0,0 +1,778 @@
+                return false;
+            }
+            $metadataProfileId = $this->targetAllMetadataFields[$metadataProfileFieldName]['metadataProfileId'];
+
+            $objectId = $dynamicObjectDetails['objectId'];
+            if (!$objectId)
+            {
+                KalturaLog::err('Invalid object ID. Cannot validate.');
+                return false;
+            }
+
+            $objects = $this->retrieveMetadataObjectsByMetadataProfileAndObjectId($metadataProfileId, KalturaMetadataObjectType::DYNAMIC_OBJECT, $objectId);
+            if ($objects->totalCount)
+            {
+                return true;
+            }
+
+            if ($dynamicObjectDetails['addIfNotExist'])
+            {
+                //retrieve metadata profile
+                $metadataProfile = $this->metadataPlugin->metadataProfile->get($metadataProfileId);
+
+                $dynamicObjectMetadataMappingProfileId = $this->getMetadataProfileId(self::OPEN_CALAIS_DYNAMIC_OBJECT_MAPPING_SYSTEM_NAME);
+                $ocmDynamicObjectMetadataMappingId = $metadataProfile->systemName . self::OPEN_CALAIS_MAPPING_POSTFIX;
+                $mappingDynamicObject = $this->retrieveMetadataObjectsByMetadataProfileAndObjectId($dynamicObjectMetadataMappingProfileId, KalturaMetadataObjectType::DYNAMIC_OBJECT, $ocmDynamicObjectMetadataMappingId);
+
+                return $this->createMissingDynamicObject($mappingDynamicObject->objects[0]->xml, $dynamicObjectDetails['fullItem'], $objectId, $metadataProfileId);
+
+            }
+
+            return false;
+        }
+
+        return true;
+    }
+
+    protected function createMissingDynamicObject($xml, $fullData, $objectId, $targetMetadataProfileId)
+    {
+        $xmlElement = new SimpleXMLElement($xml);
+        $allFields = $xmlElement->xpath('./Field');
+
+        $newObject = new SimpleXMLElement('<metadata/>');
+        foreach ($allFields as $field)
+        {
+            $sourceField = strval($field->xpath('.//Source')[0]);
+            $targetField = strval($field->xpath('.//Target')[0]);
+            $targetValue = strval($field->xpath('.//Value')[0]);
+
+            $value = null;
+            if($sourceField)
+            {
+                $value = $this->getOcmValueRecursive($fullData, explode('/', $sourceField));
+            }
+            elseif ($targetValue)
+            {
+                $value = $targetValue;
+            }
+
+            if ($value)
+            {
+                $newObject->addChild($targetField, $value);
+            }
+        }
+
+        KalturaLog::info('Creating dynamic metadata object: ' . $newObject->saveXML());
+
+        try {
+            $this->metadataPlugin->metadata->add($targetMetadataProfileId, KalturaMetadataObjectType::DYNAMIC_OBJECT, $objectId, $newObject->saveXML());
+            return $objectId;
+        }
+        catch (Exception $e)
+        {
+            KalturaLog::err("Unable to add new dynamic object with ID $objectId");
```

Comment:
Again here - better to have the batch fail than to have an error in the log that no one will see. 
Since un able to add to metadata - should be bug

Reply:
@MosheMaorKaltura  being unable to add one of several potential dynamic objects should be logged for reference, but it shouldn't crash the whole process.